### PR TITLE
AB#84951: Javascript bundling

### PR DESCRIPTION
**Pull request by AndyRae**
## Overview

Bundles the Javascript files in the core6 app, to be used in the frontend.

## Notes

- Obviously a large PR, so the main file to pay attention to is `src/submissions/api/bundleconfig.json` where the bundles are created.
- JS in `Frontend/js` is organised by pages, with ADAC renamed to Admin on a directory basis. 
- To enable IE support, the Javascript has been updated to use ES5 throughout.
- Adds the script template to the existing views in the Directory, for ease of porting over these views. No tags are added to the existing views in the core app.
- Packages not available in npm have been vendored in `Frontend/vendor`

### CODE REVIEW

**Review by dcl10 — DISMISSED** [review]

File: src/Submissions/Api/appsettings.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "ConnectionStrings": {
-    "Default": "Server=(localdb)\\MSSQLLocalDB;Database=Biobanks;Trusted_Connection=True;MultipleActiveResultSets=true",
+    "Default": "Server=localhost;Database=Biobanks;MultipleActiveResultSets=true;User ID=SA;Password=Password#1;",
```
Comment (beforan) [inline src/Submissions/Api/appsettings.json:3]: this looks like a connection string specific to your machine?

File: src/Submissions/Api/bundleconfig.json [hunk 2]
```diff
@@ -8,12 +9,585 @@
...
+    "minify": { "enabled": true },
+    "sourceMap": true
+  },
+  {
+    "outputFileName": "wwwroot/dist/js/Biobank/funders.min.js",
+    "inputFiles": [
+      "Frontend/js/shared/registerentity-funder-viewmodel.js",
+      "Frontend/js/areas/Biobank/funder-viewmodel.js",
+      "Frontend/js/areas/Biobank/funders-viewmodel.js",
+      "Frontend/node_modules/typeahead.js/dist/typeahead.bundle.min.js"
```
Comment (beforan) [inline src/Submissions/Api/bundleconfig.json:162]: typeahead is already in the layout bundle

File: src/Submissions/Api/bundleconfig.json [hunk 3]
```diff
@@ -8,12 +9,585 @@
...
+      "Frontend/js/areas/Biobank/details-logo.js",
+      "Frontend/js/areas/Biobank/edit-details.js"
+    ],
+    "minify": { "enabled": true },
+    "sourceMap": true
+  },
+  {
+    "outputFileName": "wwwroot/dist/js/Biobank/collection-and-capability-crud.min.js",
+    "inputFiles": [
+      "Frontend/node_modules/typeahead.js/dist/typeahead.bundle.min.js",
```
Comment (beforan) [inline src/Submissions/Api/bundleconfig.json:179]: typeahead is already in the layout bundle

File: src/Submissions/Api/bundleconfig.json [hunk 4]
```diff
@@ -8,12 +9,585 @@
...
+    "sourceMap": true
+  },
+  {
+    "outputFileName": "wwwroot/dist/js/Biobank/collection-sample-sets.min.js",
+    "inputFiles": [
+      "Frontend/node_modules/bootstrap-slider/dist/bootstrap-slider.min.js",
+      "Frontend/js/init/bootstrap-slider-knockout-binding.js",
+      "Frontend/js/areas/Biobank/collection-sample-sets.js"
+    ],
+    "minify": { "enabled": false },
```
Comment (beforan) [inline src/Submissions/Api/bundleconfig.json:199]: curious what the minification problem is with this one?

File: src/Submissions/Api/bundleconfig.json [hunk 5]
```diff
@@ -8,12 +9,585 @@
...
+    "outputFileName": "wwwroot/dist/js/Biobank/publications.min.js",
+    "inputFiles": ["Frontend/js/areas/Biobank/publications.js"],
+    "minify": { "enabled": true },
+    "sourceMap": true
+  },
+  {
+    "outputFileName": "wwwroot/dist/js/Biobank/collection.min.js",
+    "inputFiles": [
+      "Frontend/js/areas/Biobank/collection.js",
+      "Frontend/js/shared/associated-data-admin.js"
```
Comment (beforan) [inline src/Submissions/Api/bundleconfig.json:224]: if this file is actually shared, it shouldn't be in a per page bundle - pages which need it should include it from a shared bundle

File: src/Submissions/Api/bundleconfig.json [hunk 6]
```diff
@@ -8,12 +9,585 @@
...
+    ],
+    "minify": { "enabled": true },
+    "sourceMap": true
+  },
+  {
+    "outputFileName": "wwwroot/dist/js/Biobank/capability.min.js",
+    "inputFiles": ["Frontend/js/shared/associated-data-admin.js"],
```
Comment (beforan) [inline src/Submissions/Api/bundleconfig.json:231]: as above, make a single bundle for this file and use it on both pages

File: src/Submissions/Api/Frontend/js/areas/Biobank/submissions.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    var $btn = $(this);
+    var txtbox = $btn.data("target")
+    copyText(txtbox);
+});
+
+//generate client id
+$("#generatekey").click(function (e) {
+    var $btn = $(this);
+    $.post($btn.data("generate-url"),
+        { biobankId: $btn.data("biobank-id") }, data => {
```
Comment (beforan) [inline src/Submissions/Api/Frontend/js/areas/Biobank/submissions.js:25]: arrow functions don't work in IE. linter should complain

File: src/Submissions/Api/bundleconfig.json [hunk 8]
```diff
@@ -8,12 +9,585 @@
...
+  {
+    "outputFileName": "wwwroot/dist/js/Biobank/networkaccept.min.js",
+    "inputFiles": ["Frontend/js/areas/Biobank/network-acceptance.js"],
+    "minify": { "enabled": true },
+    "sourceMap": true
+  },
+  {
+    "outputFileName": "wwwroot/dist/js/Biobank/submissions.min.js",
+    "inputFiles": ["Frontend/js/areas/Biobank/submissions.js"],
+    "minify": { "enabled": false },
```
Comment (beforan) [inline src/Submissions/Api/bundleconfig.json:250]: curious what the minifcation issue is. this file needs es5 fixes anyway

File: src/Submissions/Api/bundleconfig.json [hunk 9]
```diff
@@ -8,12 +9,585 @@
...
+    "sourceMap": true
+  },
+  
+  //Network Admin Area
+  {
+    "outputFileName": "wwwroot/dist/js/Network/admins.min.js",
+    "inputFiles": [
+      "Frontend/js/areas/Network/admins-viewmodel.js",
+      "Frontend/js/shared/registerentity-admin-viewmodel.js"
```
Comment (beforan) [inline src/Submissions/Api/bundleconfig.json:259]: if this is shared it shouldn't be in a per page bundle; the page should include via a shared bundle instead.

File: src/Submissions/Api/bundleconfig.json [hunk 10]
```diff
@@ -8,12 +9,585 @@
...
+      "Frontend/js/areas/Biobank/edit-details.js"
+    ],
+    "minify": { "enabled": true },
+    "sourceMap": true
+  },
+  {
+    "outputFileName": "wwwroot/dist/js/Biobank/collection-and-capability-crud.min.js",
+    "inputFiles": [
+      "Frontend/node_modules/typeahead.js/dist/typeahead.bundle.min.js",
+      "Frontend/js/shared/diagnosis-type-ahead.js",
```
Comment (beforan) [inline src/Submissions/Api/bundleconfig.json:180]: if this is shared it shouldn't be in a per page bundle.

File: src/Submissions/Api/bundleconfig.json [hunk 11]
```diff
@@ -8,12 +9,585 @@
...
+  {
+    "outputFileName": "wwwroot/dist/js/Biobank/annualstats.min.js",
+    "inputFiles": ["Frontend/js/areas/Biobank/annualstats.js"],
+    "minify": { "enabled": true },
+    "sourceMap": true
+  },
+  {
+    "outputFileName": "wwwroot/dist/js/Biobank/funders.min.js",
+    "inputFiles": [
+      "Frontend/js/shared/registerentity-funder-viewmodel.js",
```
Comment (beforan) [inline src/Submissions/Api/bundleconfig.json:159]: if this is shaed it shouldn't be in a per page bundle. (if it's only used by this page, which seems likely, then move the file)

File: src/Submissions/Api/bundleconfig.json [hunk 12]
```diff
@@ -8,12 +9,585 @@
...
+    "inputFiles": ["Frontend/js/pages/Profile/publications.js"],
+    "minify": { "enabled": true },
+    "sourceMap": true
+  },
+  
+  // Biobank Area
+  {
+    "outputFileName": "wwwroot/dist/js/Biobank/admins.min.js",
+    "inputFiles": [
+      "Frontend/js/shared/registerentity-admin-viewmodel.js",
```
Comment (beforan) [inline src/Submissions/Api/bundleconfig.json:145]: this is used by more than one page, it shouldn't be in multiple bundles.

File: src/Submissions/Api/Frontend/js/areas/Admin/adac-disease-statuses.js [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+
+  this.openModalForEdit = function (_, event) {
+    _this.modal.mode(_this.modal.modalModeEdit);
+
+    // Get Row Data From DataTable
+    var rowIndex = $(event.currentTarget).data("row");
+    var diseaseStatus = dataTable.row(rowIndex).data();
+
+    let otherTerms = diseaseStatus.OtherTerms
+      ? diseaseStatus.OtherTerms.split(",").map((item) => item.trim())
```
Comment (beforan) [inline src/Submissions/Api/Frontend/js/areas/Admin/adac-disease-statuses.js:76]: arrow functions don't work in es5

File: src/Submissions/Api/Frontend/js/areas/Admin/adac-disease-statuses.js [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+  this.removeOtherTerms = function (idx) {
+    _this.modal.diseaseStatus().otherTerms.splice(idx, 1);
+  };
+  this.addAssociatedData = function () {
+    if (
+      _this.modal
+        .diseaseStatus()
+        .associatedData()
+        .find(
+          (item) => item.Id === JSON.parse($("#ass-data-select").val()).Id
```
Comment (beforan) [inline src/Submissions/Api/Frontend/js/areas/Admin/adac-disease-statuses.js:139]: arrow functions don't work in es5

File: src/Submissions/Api/Frontend/js/areas/Admin/adac-disease-statuses.js [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+          render: function (data, type, row, meta) {
+
+              let returnString = "N/A";
+          if (data) {
+              if (data.length === 1) {
+                  returnString = `<li>${data[0].Name}</li>`;
+              } else if (data.length < 4 && data.length > 1) {
+                  // return their values as a list
+                  let displayData = data
+                      .map((item) => `<li>${item.Name}</li>`)
```
Comment (beforan) [inline src/Submissions/Api/Frontend/js/areas/Admin/adac-disease-statuses.js:211]: no es5 arrow functions

Comment (beforan) [reply src/Submissions/Api/Frontend/js/areas/Admin/adac-disease-statuses.js:211]: or template strings

File: src/Submissions/Api/bundleconfig.json [hunk 16]
```diff
@@ -8,12 +9,585 @@
...
+    "inputFiles": ["Frontend/js/areas/Network/network-biobanks.js"],
+    "minify": { "enabled": true },
+    "sourceMap": true
+  },
+  
+  // //ADAC Admin Area
+  {
+    "outputFileName": "wwwroot/dist/js/Admin/disease-statuses.min.js",
+    "inputFiles": ["Frontend/js/areas/Admin/adac-disease-statuses.js"],
+    "minify": { "enabled": false },
```
Comment (beforan) [inline src/Submissions/Api/bundleconfig.json:290]: curious on the issue here? However this file has a lot of es5 issues (arrow functions, template strings). The linter should catch them.

File: src/Submissions/Api/bundleconfig.json [hunk 17]
```diff
@@ -8,12 +9,585 @@
...
+    "sourceMap": true
+  },
+  {
+    "outputFileName": "wwwroot/dist/js/Admin/material-types.min.js",
+    "inputFiles": ["Frontend/js/areas/Admin/adac-material-types.js"],
+    "minify": { "enabled": false },
```
Comment (beforan) [inline src/Submissions/Api/bundleconfig.json:296]: minification issues? didn't spot any es5 issues in this file (at a glance) so curious.

File: src/Submissions/Api/bundleconfig.json [hunk 18]
```diff
@@ -8,12 +9,585 @@
...
+  {
+    "outputFileName": "wwwroot/dist/js/Admin/material-type-groups.min.js",
+    "inputFiles": ["Frontend/js/areas/Admin/adac-material-type-groups.js"],
+    "minify": { "enabled": true },
+    "sourceMap": true
+  },
+  {
+    "outputFileName": "wwwroot/dist/js/Admin/sexes.min.js",
+    "inputFiles": ["Frontend/js/areas/Admin/adac-sexes.js"],
+    "minify": { "enabled": false },
```
Comment (beforan) [inline src/Submissions/Api/bundleconfig.json:308]: didn't spot es5 issues in here, so what's the minification issue?

File: src/Submissions/Api/bundleconfig.json [hunk 19]
```diff
@@ -8,12 +9,585 @@
...
+    "sourceMap": true
+  },
+  {
+    "outputFileName": "wwwroot/dist/js/Admin/country.min.js",
+    "inputFiles": ["Frontend/js/areas/Admin/adac-country.js"],
+    "minify": { "enabled": false },
```
Comment (beforan) [inline src/Submissions/Api/bundleconfig.json:314]: same again, what's the issue as I can't spot it?

**Review by beforan — CHANGES_REQUESTED** [review]
My comments aren't exhaustive, but there are a couple of common themes to address throughout.

1. The minifier not supporting `??` isn't a problem in and of itself; not minifying is a sensible fix
  - however in practice for us it exposes a real issue: IE support
  - There's a reason the eslint config this PR adds specifies es5. All files should (manually because we don't have babel or similar in this project) be es5 so that they work in IE.
  - This means no arrow functions, no const, no let, no template strings and no nullish coalescing/propragation etc.
  - The code obviously previously should have adhered to that, but sometimes new code introduced broke it
  - Now we actually have linter config / js project code managed ina. somewhat modern way, we can better keep an eye on this, so this PR is an opportunity to fix these.
2. files appearing in more than one bundle.
  - I don't think in any case we should double-bundle source if possible. If two pages need a file, both pages should include that file's bundle, as opposed to the file being present in both page's bundles.
  - shared files should be bundled as high up the chain as makes sense. If they're light, and "safe" they should be in the layout bundle. If they're heavy, or unsafe, they should be in a bundle specific to a set of pages only (e.g. the markdown editor). If they're only used on a single page, they can go into a page bundle.
3. The source uses both `pages/` and `areas/` paths, but they seem to serve the same purpose from the perspective of logical organisation (e.g. even the `pages/` files are actually grouped by area/controller anyway); please consolidate into one or the other.

In summary:
- fix es5 errors throughout
- enable minifying hopefully in light of that
- don't double bundle files - any single source file should only be in one bundle
- shared files should be bundled as broadly as is sensible (balancing need vs bundle size)
- eliminate either `pages/` or `areas/`

File: src/Submissions/Api/bundleconfig.json (hunk 4 above)
Comment (AndyRae) [reply src/Submissions/Api/bundleconfig.json:201]: Line 516 in `collection-sample-sets.js` for custom knockout validation rules:

```
$(function () {
  ko.validation.rules["mustEqual"] = {
	validator(val, otherVal) {
	  return val === otherVal;
	},
}
```

Should be:
```
$(function () {
  ko.validation.rules["mustEqual"] = {
	validator: function(val, otherVal) {
	  return val === otherVal;
	},
}
```

File: src/Submissions/Api/appsettings.json (hunk 1 above)
Comment (AndyRae) [reply src/Submissions/Api/appsettings.json]: Removed

File: src/Submissions/Api/bundleconfig.json (hunk 2 above)
Comment (AndyRae) [reply src/Submissions/Api/bundleconfig.json]: Removed

File: src/Submissions/Api/bundleconfig.json (hunk 5 above)
Comment (AndyRae) [reply src/Submissions/Api/bundleconfig.json]: Moved to a shared bundle.

File: src/Submissions/Api/bundleconfig.json (hunk 6 above)
Comment (AndyRae) [reply src/Submissions/Api/bundleconfig.json:230]: Moved to a shared bundle.

File: src/Submissions/Api/bundleconfig.json (hunk 8 above)
Comment (AndyRae) [reply src/Submissions/Api/bundleconfig.json]: The problem is actually in producing the source map, and not minifying. So have turned off the source map for the bundles that produce errors, and turned minify back on.

File: src/Submissions/Api/Frontend/js/areas/Admin/adac-disease-statuses.js (hunk 15 above)
Comment (AndyRae) [reply src/Submissions/Api/Frontend/js/areas/Admin/adac-disease-statuses.js]: Updated

File: src/Submissions/Api/Frontend/js/areas/Admin/adac-disease-statuses.js (hunk 14 above)
Comment (AndyRae) [reply src/Submissions/Api/Frontend/js/areas/Admin/adac-disease-statuses.js]: Updated

File: src/Submissions/Api/Frontend/js/areas/Admin/adac-disease-statuses.js (hunk 13 above)
Comment (AndyRae) [reply src/Submissions/Api/Frontend/js/areas/Admin/adac-disease-statuses.js]: Updated

File: src/Submissions/Api/bundleconfig.json (hunk 12 above)
Comment (AndyRae) [reply src/Submissions/Api/bundleconfig.json]: Updated - moved to its own bundle.

File: src/Submissions/Api/bundleconfig.json (hunk 11 above)
Comment (AndyRae) [reply src/Submissions/Api/bundleconfig.json]: Updated the bundle.

File: src/Submissions/Api/bundleconfig.json (hunk 10 above)
Comment (AndyRae) [reply src/Submissions/Api/bundleconfig.json]: Updated - was only needed as a page bundle so moved the file.

File: src/Submissions/Api/bundleconfig.json (hunk 9 above)
Comment (AndyRae) [reply src/Submissions/Api/bundleconfig.json]: Updated - moved to a shared bundle.

File: src/Submissions/Api/Frontend/js/areas/Biobank/submissions.js (hunk 7 above)
Comment (AndyRae) [reply src/Submissions/Api/Frontend/js/areas/Biobank/submissions.js]: Updated to ES5

**Review by beforan — APPROVED** [review]
